File: tools/tour/evidence-validator.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# Two independent layers of authentication, neither of which trusts the
# producer:
#
#  1. STRUCTURAL — every field is re-derived from data the validator itself
#     reads (inventory, accepted baseline, pin files, normalizer) and must
#     match what the ledger claims. This alone cannot detect a ledger whose
#     raw bytes were simply copied between rows: a copy is internally
#     consistent, so every recomputed hash/root/verdict still balances.
#  2. REPLAY (the "process" phase) — the validator re-executes every
#     attempt's command against a freshly materialized module, using the
#     exact pinned Go 1.27 binary and the exact bashy binary named in the
#     manifest, and requires the fresh spawn/state/exit and (for the two
#     Bash++ modes) the fresh normalized output to match what the ledger
#     recorded. This is the external attestation: a forged row's stored
#     bytes did not come from executing ITS OWN bound command, so replaying
#     that command reproduces different bytes and the row is rejected — no
#     matter how consistently the rest of the ledger was recomputed around
#     it. Pinned-Go baseline rows are the one place byte drift is tolerated
#     (official Go programs are not all deterministic), but only once
#     spawn/state/exit already match the pinned toolchain's fresh run.

require_relative 'evidence'

root = File.expand_path('../..', __dir__)
file = ENV.fetch('TOUR_EVIDENCE', File.join(root, 'tests/tour/evidence.jsonl'))
normalizer = ENV.fetch('TOUR_NORMALIZER', File.join(root, 'tools/tour/normalize.rb'))
abort "FATAL: missing evidence #{file}" unless File.file?(file)
records = File.readlines(file, chomp: true).map.with_index do |line, i|
  record = JSON.parse(line)
  abort "FATAL: non-canonical JSON line #{i + 1}" unless line == TourEvidence.canonical(record)
  record
end
manifest, verdict = records.first, records.last
abort 'FATAL: malformed evidence envelope' unless manifest['type'] == 'manifest' && records[-3]['type'] == 'summary' && records[-2]['type'] == 'root' && verdict['type'] == 'verdict'
abort 'FATAL: schema mismatch' unless manifest['schema'] == TourEvidence::SCHEMA
abort 'FATAL: normalizer checksum mismatch' unless manifest.dig('normalizer', 'sha256') == TourEvidence.sha(File.binread(normalizer))

inventory_file = File.join(root, manifest.dig('inventory', 'path'))
inventory_lines = File.readlines(inventory_file, chomp: true).reject { |l| l.start_with?('#') || l.empty? }
inventory = inventory_lines.map do |line|
  f = line.split("\t"); next unless %w[applicable_go_program build_only_go_program].include?(f[3])
  [f[0], { 'path' => f[0], 'applicability' => f[3], 'bytes' => Integer(f[6]), 'sha256' => f[7] }]
end.compact.to_h
data_sha = TourEvidence.sha((inventory_lines.join("\n") + "\n"))
abort 'FATAL: manifest inventory binding mismatch' unless inventory.length == 97 && manifest.dig('inventory', 'executable_programs') == 97 && manifest.dig('inventory', 'data_sha256') == data_sha

base_file = File.join(root, manifest.dig('baseline', 'accepted_results'))
pin_file = File.join(root, manifest.dig('baseline', 'pin'))
abort 'FATAL: accepted baseline file binding mismatch' unless TourEvidence.sha(File.binread(base_file)) == manifest.dig('baseline', 'accepted_results_sha256')
abort 'FATAL: baseline pin binding mismatch' unless TourEvidence.sha(File.binread(pin_file)) == manifest.dig('baseline', 'pin_sha256')
unless system({ 'TOUR_RESULTS' => base_file }, File.join(root, 'tools/tour/validate-results.sh'), out: File::NULL)
  abort 'FATAL: accepted baseline observations fail their independent validator'
end
baseline_ledger = File.readlines(base_file, chomp: true).map { |l| next if l.start_with?('#') || l.empty?; f = l.split("\t"); [f[0], f] }.compact.to_h

# --- exact Go 1.27 builder: cross-checked against the pin file, not just
# self-consistency inside the manifest (a manifest could pair a fabricated
# identity with a matching fabricated checksum for some OTHER real binary,
# e.g. a Go 1.26 toolchain, and self-consistency alone would not catch it).
pin_row = File.readlines(File.join(root, 'docs/tour/pin.tsv'), chomp: true).find { |l| !l.start_with?('#') && !l.empty? }.split("\t")
tour_version = pin_row[1]
helper_row = File.readlines(File.join(root, 'docs/tour/helpers.tsv'), chomp: true).find { |l| !l.start_with?('#') && !l.empty? }.split("\t")
goos = `uname -s`.strip.downcase
goarch = `uname -m`.strip
tc_row = File.readlines(File.join(root, 'docs/tour/toolchain.tsv'), chomp: true)
  .reject { |l| l.start_with?('#') || l.empty? }
  .map { |l| l.split("\t") }
  .find { |f| f[0] == goos && f[1] == goarch }
abort "FATAL: no pinned Go toolchain row for #{goos}/#{goarch} in docs/tour/toolchain.tsv" unless tc_row
tc_version, tc_identity, tc_sha = tc_row[2], tc_row[3], tc_row[4]

go = manifest['go']
abort 'FATAL: exact Go binary unavailable' unless go['present'] && File.executable?(go['path'])
abort 'FATAL: exact Go binary checksum mismatch' unless TourEvidence.sha(File.binread(go['path'])) == go['sha256']
abort 'FATAL: exact Go identity mismatch' unless `#{go['path']} version`.strip == go['identity']
abort "FATAL: Go builder is not the pinned Go 1.27 toolchain (manifest identity #{go['identity']}, pinned #{tc_identity})" unless go['identity'] == tc_identity
abort "FATAL: Go builder checksum is not the pinned Go 1.27 toolchain (manifest #{go['sha256']}, pinned #{tc_sha})" unless go['sha256'] == tc_sha && go['pinned_sha256'] == tc_sha

# --- clean published reproducible bashy: the version string is re-derived
# from the LIVE binary (exactly like the Go identity above) and the published/
# clean/dirty verdict is parsed out of that live string — never taken on the
# manifest's say-so. Otherwise a ledger could pair a dev/dirty binary's true
# checksum with a fabricated published version string: every self-consistency
# check would balance while the evidence claims a release that was never built.
bashy = manifest['bashy']
abort 'FATAL: bashy binary unavailable' unless bashy['present'] && File.executable?(bashy['path'])
abort 'FATAL: exact bashy executable checksum mismatch' unless TourEvidence.sha(File.binread(bashy['path'])) == bashy['sha256']
live_bashy_version = `#{bashy['path']} --version 2>/dev/null`.lines.first.to_s.strip
abort "FATAL: exact bashy version mismatch — manifest claims #{bashy['version'].inspect}, live binary reports #{live_bashy_version.inspect}" unless live_bashy_version == bashy['version']
identity = TourEvidence.bashy_identity(live_bashy_version)
abort "FATAL: bashy build is dirty or unpublished (#{bashy['version'].inspect}) — a workspace/dirty build is not clean published reproducible evidence" unless identity['published'] && !identity['dirty']
abort 'FATAL: bashy source_revision does not match its own version string' unless bashy['source_revision'] == identity['revision']
abort 'FATAL: bashy published flag does not match its own version string' unless bashy['published'] == true
abort 'FATAL: bashy reproducible flag does not match its own version string' unless bashy['reproducible'] == true

attempts = records[1...-3]
abort "FATAL: coverage must be exactly 97 x 3 = 291" unless attempts.length == 291 && manifest['attempts'] == 291
seen = {}; baselines = {}; recomputed_counts = Hash.new(0)
attempts.each do |attempt|
  path, mode = attempt.values_at('path', 'mode'); key = [path, mode]
  abort "FATAL: unexpected or duplicate attempt #{path} #{mode}" unless inventory.key?(path) && TourEvidence::MODES.include?(mode) && !seen[key]
  seen[key] = true; item = inventory.fetch(path)
  expected_source = { 'bytes' => item['bytes'], 'sha256' => item['sha256'] }
  abort "FATAL: inventory/source mismatch #{path} #{mode}" unless attempt['applicability'] == item['applicability'] && attempt['source'] == expected_source
  abort "FATAL: unknown pipeline stage #{path} #{mode}" unless TourEvidence::STAGES.include?(attempt['stage'])
  command = attempt['command']
  if mode != 'baseline' && baselines[path] && command == baselines[path]['command']
    abort "FATAL: synthetic mutual-equality artifact #{path} #{mode}"
  end
  case mode
  when 'baseline'
    abort "FATAL: unbound command #{path} #{mode}" unless command.is_a?(Array) && !command.empty?
    abort "FATAL: baseline did not use exact Go #{path}" unless File.expand_path(command[0]) == File.expand_path(go['path'])
    abort "FATAL: baseline stage must be run #{path}" unless attempt['stage'] == 'run'
  when 'interpreted'
    abort "FATAL: unbound command #{path} #{mode}" unless command.is_a?(Array) && !command.empty?
    abort "FATAL: interpreted command binding #{path}" unless command[0] == bashy['path'] && command.include?('--bashpp')
    abort "FATAL: interpreted stage must be run #{path}" unless attempt['stage'] == 'run'
  when 'compiled'
    abort "FATAL: unbound compiled pipeline #{path}" unless command.is_a?(Hash) && command.keys.sort == %w[build run transpile]
    abort "FATAL: compiled transpile binding #{path}" unless command['transpile'][0] == bashy['path'] && command['transpile'].include?('transpile')
    abort "FATAL: compiled build binding #{path}" unless command['build'][0] == go['path'] && command['build'].include?('build')
    abort "FATAL: compiled run binding #{path}" unless command['run'].is_a?(Array) && command['run'].length == 1
  end
  raw = { 'stdout' => Base64.strict_decode64(attempt.dig('raw', 'stdout_base64')), 'stderr' => Base64.strict_decode64(attempt.dig('raw', 'stderr_base64')) }
  abort "FATAL: agent-specific Bashy advertisement in evidence #{path} #{mode}" if TourEvidence.agent_banner?(raw)
  derived = TourEvidence.derived(raw, normalizer)
  abort "FATAL: normalized fields not derived from raw streams #{path} #{mode}" unless attempt['normalized'] == derived
  if mode == 'baseline'
    row = baseline_ledger.fetch(path); build_only = item['applicability'] == 'build_only_go_program'
    expected = { 'exit' => build_only ? Integer(row[6]) : Integer(row[7]),
      'stdout_bytes' => build_only ? nil : Integer(row[8]), 'stdout_sha256' => build_only ? nil : row[9],
      'stderr_bytes' => build_only ? nil : Integer(row[10]), 'stderr_sha256' => build_only ? nil : row[11] }
    abort "FATAL: baseline accepted observation copied or altered #{path}" unless attempt['accepted_observation'] == expected
    authentic = attempt['exit'] == expected['exit'] && (build_only ? raw['stdout'].empty? && raw['stderr'].empty? :
      derived['stdout']['bytes'] == expected['stdout_bytes'] && derived['stdout']['sha256'] == expected['stdout_sha256'] &&
      derived['stderr']['bytes'] == expected['stderr_bytes'] && derived['stderr']['sha256'] == expected['stderr_sha256'])
    abort "FATAL: accepted baseline match flag is not derived #{path}" unless attempt['accepted_observation_matches'] == authentic
    # Volatile official examples need not reproduce a prior stream byte for
    # byte. Authenticity comes from the validated accepted ledger, exact Go
    # binary, pinned source, bound command, captured process facts, AND (see
    # the replay phase below) an independent live re-execution; this flag
    # only records whether the fresh observation also equals the PRIOR one.
    calculated = TourEvidence.outcome(attempt)
    baselines[path] = attempt
  else
    abort "FATAL: mode precedes its baseline #{path}" unless baselines[path]
    calculated = TourEvidence.outcome(attempt, baselines[path])
    # Equality is accepted only when it came through a distinct, correctly
    # bound execution record. This rejects fabricated baseline clones.
  end
  abort "FATAL: derived outcome mismatch #{path} #{mode}" unless attempt['outcome'] == calculated
  recomputed_counts[calculated] += 1
end
inventory.each_key { |path| TourEvidence::MODES.each { |mode| abort "FATAL: missing attempt #{path} #{mode}" unless seen[[path, mode]] } }
summary = records[-3]
expected_summary = { 'type' => 'summary', 'attempts' => 291, 'programs' => 97, 'modes' => TourEvidence::MODES, 'outcomes' => recomputed_counts.sort.to_h }
abort 'FATAL: summary mismatch' unless summary == expected_summary
calculated_root = TourEvidence.root(records[0...-2])
abort 'FATAL: evidence root mismatch' unless records[-2] == { 'type' => 'root', 'algorithm' => 'sha256-canonical-jsonl', 'sha256' => calculated_root }
expected_verdict = recomputed_counts == { 'PASS' => 291 } ? 'PASS' : 'FAIL'
abort 'FATAL: verdict mismatch' unless verdict == { 'type' => 'verdict', 'value' => expected_verdict, 'root_sha256' => calculated_root }

# ---------------------------------------------------------------------------
# PROCESS PHASE — replay authentication.
#
# Everything above is self-consistency: it proves the ledger agrees with
# itself and with the files it cites. None of it proves the raw bytes in any
# given row actually came from executing THAT row's own bound command — a
# row whose raw/normalized/spawn/state/exit fields were copied from a
# DIFFERENT row (e.g. its sibling baseline attempt) while keeping its own
# correct command is internally perfect and would sail through every check
# above. Replay closes that gap by re-executing every attempt for real,
# right now, against a freshly materialized module, and requiring the fresh
# observation to match the recorded one.
# ---------------------------------------------------------------------------
timeout = Integer(ENV.fetch('TOUR_STEP_TIMEOUT', '30'))
tour_root = ENV.fetch('TOUR_ROOT', File.join(`go env GOMODCACHE`.strip, "golang.org/x/website@#{tour_version}"))
Dir.mktmpdir('tour-evidence-replay') do |work|
  mod = File.join(work, 'module')
  begin
    TourEvidence.materialize_module(mod, tour_root: tour_root, inventory: inventory.values, go_version: tc_version, helper: helper_row)
  rescue StandardError => e
    abort "FATAL: cannot materialize replay module: #{e.message}"
  end

  attempts.each do |attempt|
    path, mode = attempt.values_at('path', 'mode')
    item = inventory.fetch(path)
    local = File.join(mod, path)
    env = { 'GOTOOLCHAIN' => 'local', 'BASHY_HINTS' => 'off' }
    fresh =
      if mode == 'compiled'
        compiled_dir = File.join(work, 'compiled', path.gsub(%r{[^\w./]}, '_'))
        FileUtils.mkdir_p(compiled_dir)
        pipeline = TourEvidence.compiled_pipeline(bashy: bashy['path'], go_bin: go['path'], source: local, module_dir: mod, workdir: compiled_dir, timeout: timeout, env: env)
        abort "FATAL: replay pipeline stage mismatch #{path} #{mode} (recorded #{attempt['stage']}, replay #{pipeline['stage']})" unless pipeline['stage'] == attempt['stage']
        pipeline['raw']
      else
        command = TourEvidence.command_for(mode, item, bashy: bashy['path'], go_bin: go['path'], local: local)
        TourEvidence.capture(command, chdir: mod, timeout: timeout, env: env)
      end

    fresh_normalized = TourEvidence.derived(fresh, normalizer)
    control_match = fresh['spawned'] == attempt['spawned'] && fresh['state'] == attempt['state'] && fresh['exit'] == attempt['exit']
    abort "FATAL: replay spawn/state/exit not reproduced #{path} #{mode} (recorded exit=#{attempt['exit']} state=#{attempt['state']}, replay exit=#{fresh['exit']} state=#{fresh['state']})" unless control_match

    bytes_match = %w[stdout stderr].all? { |stream| fresh_normalized[stream]['sha256'] == attempt.dig('normalized', stream, 'sha256') }
    next if bytes_match
    # Pinned-Go baseline rows are the one place byte drift is tolerated: not
    # every official Go program is deterministic, and this project's own
    # accepted-baseline design already accounts for that (see
    # accepted_observation_matches). The two Bash++ modes are held to exact
    # replay equality: their whole purpose is to be the authoritative record
    # of what bashy did, and a mismatch there means the recorded bytes did
    # not come from this row's own bound command.
    next if mode == 'baseline'
    abort "FATAL: replay raw-output mismatch #{path} #{mode} — recorded observation was not reproduced by executing its own bound command"
  end
end

puts "Tour evidence VALID #{expected_verdict}: 97 programs x 3 modes = 291; root #{calculated_root}; replay-authenticated"

File: tools/tour/evidence_validator.go
// Sprint 155 / Story S155.9 / Story-ID 43af37063b09.
//
// tour-evidence/v2 validator — the port of tools/tour/evidence-validator.rb.
// Two independent layers, neither of which trusts the producer:
//
//  1. STRUCTURAL — every field is re-derived from data the validator itself
//     reads (inventory, accepted baseline, pin files, normalizer) and must
//     match what the ledger claims.
//  2. REPLAY — every attempt's command is re-executed against a freshly
//     materialized module with the exact pinned Go binary and the exact bashy
//     binary named in the manifest; the fresh spawn/state/exit and (for the
//     two Bash++ modes) the fresh normalized output must match the record.
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

func cmdEvidenceValidator(root string) int {
	file := envOr("TOUR_EVIDENCE", filepath.Join(root, "tests/tour/evidence.jsonl"))
	normalizer := envOr("TOUR_NORMALIZER", filepath.Join(root, normalizerPath))
	if !fileExists(file) {
		return abortf("FATAL: missing evidence %s", file)
	}
	records, err := readLedger(file)
	if err != nil {
		if strings.HasPrefix(err.Error(), "non-canonical") {
			return abortf("FATAL: %s", strings.Replace(err.Error(), "non-canonical JSON on line", "non-canonical JSON line", 1))
		}
		return abortf("FATAL: %v", err)
	}
	if len(records) < 4 {
		return abortf("FATAL: malformed evidence envelope")
	}
	manifest, verdict := records[0], records[len(records)-1]
	if !(manifest["type"] == "manifest" && records[len(records)-3]["type"] == "summary" && records[len(records)-2]["type"] == "root" && verdict["type"] == "verdict") {
		return abortf("FATAL: malformed evidence envelope")
	}
	if manifest["schema"] != evidenceSchema {
		return abortf("FATAL: schema mismatch")
	}
	// The normalizer binding: the on-disk Go implementation, or the retired
	// Ruby script's frozen digest for a ledger sealed before the port.
	normalizerOK := jsonEqual(dig(manifest, "normalizer", "sha256"), shaFile(normalizer)) ||
		(dig(manifest, "normalizer", "path") == retiredRubyNormalizerPath && jsonEqual(dig(manifest, "normalizer", "sha256"), retiredRubyNormalizerSHA256))
	if !normalizerOK {
		return abortf("FATAL: normalizer checksum mismatch")
	}

	inventoryFile := filepath.Join(root, toS(dig(manifest, "inventory", "path")))
	items, dataSHA := evidenceInventory(inventoryFile)
	inventory := map[string]Item{}
	for _, item := range items {
		inventory[item.Path] = item
	}
	if !(len(inventory) == 97 && jsonEqual(dig(manifest, "inventory", "executable_programs"), int64(97)) && dig(manifest, "inventory", "data_sha256") == dataSHA) {
		return abortf("FATAL: manifest inventory binding mismatch")
	}

	baseFile := filepath.Join(root, toS(dig(manifest, "baseline", "accepted_results")))
	pinFile := filepath.Join(root, toS(dig(manifest, "baseline", "pin")))
	if !jsonEqual(shaFile(baseFile), dig(manifest, "baseline", "accepted_results_sha256")) {
		return abortf("FATAL: accepted baseline file binding mismatch")
	}
	if !jsonEqual(shaFile(pinFile), dig(manifest, "baseline", "pin_sha256")) {
		return abortf("FATAL: baseline pin binding mismatch")
	}
	validate := exec.Command(filepath.Join(root, "tools/tour/validate-results.sh"))
	validate.Env = append(os.Environ(), "TOUR_RESULTS="+baseFile)
	validate.Stdout = nil
	validate.Stderr = os.Stderr
	if err := validate.Run(); err != nil {
		return abortf("FATAL: accepted baseline observations fail their independent validator")
	}
	baselineLedger := map[string][]string{}
	for _, f := range tsvRowsLoose(baseFile) {
		baselineLedger[field(f, 0)] = f
	}

	// --- exact Go 1.27 builder: cross-checked against the pin file
	pinRow := firstDataRow(filepath.Join(root, "docs/tour/pin.tsv"))
	tourVersion := pinRow[1]
	helperRow := firstDataRow(filepath.Join(root, "docs/tour/helpers.tsv"))
	goos, goarch := hostGoosGoarch()
	var tcRow []string
	for _, f := range tsvRowsLoose(filepath.Join(root, "docs/tour/toolchain.tsv")) {
		if field(f, 0) == goos && field(f, 1) == goarch {
			tcRow = f
			break
		}
	}
	if tcRow == nil {
		return abortf("FATAL: no pinned Go toolchain row for %s/%s in docs/tour/toolchain.tsv", goos, goarch)
	}
	tcVersion, tcIdentity, tcSHA := tcRow[2], tcRow[3], tcRow[4]

	goInfo := asMap(manifest["go"])
	goPath := toS(goInfo["path"])
	if !(truthy(goInfo["present"]) && isExecutable(goPath)) {
		return abortf("FATAL: exact Go binary unavailable")
	}
	if shaFile(goPath) != toS(goInfo["sha256"]) {
		return abortf("FATAL: exact Go binary checksum mismatch")
	}
	if shellOutput(nil, goPath, "version") != toS(goInfo["identity"]) {
		return abortf("FATAL: exact Go identity mismatch")
	}
	if toS(goInfo["identity"]) != tcIdentity {
		return abortf("FATAL: Go builder is not the pinned Go 1.27 toolchain (manifest identity %s, pinned %s)", toS(goInfo["identity"]), tcIdentity)
	}
	if !(toS(goInfo["sha256"]) == tcSHA && toS(goInfo["pinned_sha256"]) == tcSHA) {
		return abortf("FATAL: Go builder checksum is not the pinned Go 1.27 toolchain (manifest %s, pinned %s)", toS(goInfo["sha256"]), tcSHA)
	}

	// --- clean published reproducible bashy: re-derived from the LIVE binary
	bashy := asMap(manifest["bashy"])
	bashyPath := toS(bashy["path"])
	if !(truthy(bashy["present"]) && isExecutable(bashyPath)) {
		return abortf("FATAL: bashy binary unavailable")
	}
	if shaFile(bashyPath) != toS(bashy["sha256"]) {
		return abortf("FATAL: exact bashy executable checksum mismatch")
	}
	liveOut, _ := exec.Command(bashyPath, "--version").Output()
	liveBashyVersion := strings.TrimSpace(firstLine(string(liveOut)))
	if liveBashyVersion != toS(bashy["version"]) {
		return abortf("FATAL: exact bashy version mismatch — manifest claims %s, live binary reports %s", inspectString(toS(bashy["version"])), inspectString(liveBashyVersion))
	}
	identity := bashyIdentity(liveBashyVersion)
	if !(truthy(identity["published"]) && !truthy(identity["dirty"])) {
		return abortf("FATAL: bashy build is dirty or unpublished (%s) — a workspace/dirty build is not clean published reproducible evidence", inspectString(toS(bashy["version"])))
	}
	if !jsonEqual(bashy["source_revision"], identity["revision"]) {
		return abortf("FATAL: bashy source_revision does not match its own version string")
	}
	if bashy["published"] != true {
		return abortf("FATAL: bashy published flag does not match its own version string")
	}
	if bashy["reproducible"] != true {
		return abortf("FATAL: bashy reproducible flag does not match its own version string")
	}

	attempts := records[1 : len(records)-3]
	if !(len(attempts) == 291 && jsonEqual(manifest["attempts"], int64(291))) {
		return abortf("FATAL: coverage must be exactly 97 x 3 = 291")
	}
	type key struct{ path, mode string }
	seen := map[key]bool{}
	baselines := map[string]map[string]any{}
	recomputedCounts := map[string]any{}
	for _, attempt := range attempts {
		path, mode := asString(attempt["path"]), asString(attempt["mode"])
		k := key{path, mode}
		item, known := inventory[path]
		if !(known && containsString(modes, mode) && !seen[k]) {
			return abortf("FATAL: unexpected or duplicate attempt %s %s", path, mode)
		}
		seen[k] = true
		expectedSource := map[string]any{"bytes": item.Bytes, "sha256": item.SHA256}
		if !(attempt["applicability"] == item.Applicability && jsonEqual(attempt["source"], expectedSource)) {
			return abortf("FATAL: inventory/source mismatch %s %s", path, mode)
		}
		if !containsString(evidenceStages, asString(attempt["stage"])) {
			return abortf("FATAL: unknown pipeline stage %s %s", path, mode)
		}
		command := attempt["command"]
		if mode != "baseline" && baselines[path] != nil && jsonEqual(command, baselines[path]["command"]) {
			return abortf("FATAL: synthetic mutual-equality artifact %s %s", path, mode)
		}
		switch mode {
		case "baseline":
			argv, isList := command.([]any)
			if !isList || len(argv) == 0 {
				return abortf("FATAL: unbound command %s %s", path, mode)
			}
			if expandPath(asString(argv[0])) != expandPath(goPath) {
				return abortf("FATAL: baseline did not use exact Go %s", path)
			}
			if attempt["stage"] != "run" {
				return abortf("FATAL: baseline stage must be run %s", path)
			}
		case "interpreted":
			argv, isList := command.([]any)
			if !isList || len(argv) == 0 {
				return abortf("FATAL: unbound command %s %s", path, mode)
			}
			if !(asString(argv[0]) == bashyPath && containsString(strList(argv), "--bashpp")) {
				return abortf("FATAL: interpreted command binding %s", path)
			}
			if attempt["stage"] != "run" {
				return abortf("FATAL: interpreted stage must be run %s", path)
			}
		case "compiled":
			pipeline, isMap := command.(map[string]any)
			if !isMap || !equalStrings(sortedKeys(pipeline), []string{"build", "run", "transpile"}) {
				return abortf("FATAL: unbound compiled pipeline %s", path)
			}
			transpile := strList(pipeline["transpile"])
			if !(len(transpile) > 0 && transpile[0] == bashyPath && containsString(transpile, "transpile")) {
				return abortf("FATAL: compiled transpile binding %s", path)
			}
			build := strList(pipeline["build"])
			if !(len(build) > 0 && build[0] == goPath && containsString(build, "build")) {
				return abortf("FATAL: compiled build binding %s", path)
			}
			runArgv, isRunList := pipeline["run"].([]any)
			if !isRunList || len(runArgv) != 1 {
				return abortf("FATAL: compiled run binding %s", path)
			}
		}
		stdout, okOut := strictB64(toS(dig(attempt, "raw", "stdout_base64")))
		stderr, okErr := strictB64(toS(dig(attempt, "raw", "stderr_base64")))
		if !okOut || !okErr {
			return abortf("FATAL: invalid base64 in raw streams %s %s", path, mode)
		}
		if agentBanner(stderr) {
			return abortf("FATAL: agent-specific Bashy advertisement in evidence %s %s", path, mode)
		}
		derived := evidenceDerived(stdout, stderr)
		if !jsonEqual(attempt["normalized"], derived) {
			return abortf("FATAL: normalized fields not derived from raw streams %s %s", path, mode)
		}
		var calculated string
		if mode == "baseline" {
			row := baselineLedger[path]
			buildOnly := item.Applicability == "build_only_go_program"
			expected := expectedAccepted(row, buildOnly)
			if !jsonEqual(attempt["accepted_observation"], expected) {
				return abortf("FATAL: baseline accepted observation copied or altered %s", path)
			}
			authentic := jsonEqual(attempt["exit"], expected["exit"])
			if buildOnly {
				authentic = authentic && len(stdout) == 0 && len(stderr) == 0
			} else {
				authentic = authentic && jsonEqual(dig(derived, "stdout", "bytes"), expected["stdout_bytes"]) &&
					jsonEqual(dig(derived, "stdout", "sha256"), expected["stdout_sha256"]) &&
					jsonEqual(dig(derived, "stderr", "bytes"), expected["stderr_bytes"]) &&
					jsonEqual(dig(derived, "stderr", "sha256"), expected["stderr_sha256"])
			}
			if !jsonEqual(attempt["accepted_observation_matches"], authentic) {
				return abortf("FATAL: accepted baseline match flag is not derived %s", path)
			}
			calculated = evidenceOutcome(attempt, nil)
			baselines[path] = attempt
		} else {
			if baselines[path] == nil {
				return abortf("FATAL: mode precedes its baseline %s", path)
			}
			calculated = evidenceOutcome(attempt, baselines[path])
		}
		if attempt["outcome"] != calculated {
			return abortf("FATAL: derived outcome mismatch %s %s", path, mode)
		}
		recomputedCounts[calculated] = toI(recomputedCounts[calculated]) + 1
	}
	for path := range inventory {
		for _, mode := range modes {
			if !seen[key{path, mode}] {
				return abortf("FATAL: missing attempt %s %s", path, mode)
			}
		}
	}
	summary := records[len(records)-3]
	expectedSummary := map[string]any{"type": "summary", "attempts": int64(291), "programs": int64(97), "modes": anyList(modes), "outcomes": recomputedCounts}
	if !jsonEqual(summary, expectedSummary) {
		return abortf("FATAL: summary mismatch")
	}
	calculatedRoot := ledgerRoot(records[:len(records)-2])
	if !jsonEqual(records[len(records)-2], map[string]any{"type": "root", "algorithm": "sha256-canonical-jsonl", "sha256": calculatedRoot}) {
		return abortf("FATAL: evidence root mismatch")
	}
	expectedVerdict := "FAIL"
	if jsonEqual(recomputedCounts, map[string]any{"PASS": int64(291)}) {
		expectedVerdict = "PASS"
	}
	if !jsonEqual(verdict, map[string]any{"type": "verdict", "value": expectedVerdict, "root_sha256": calculatedRoot}) {
		return abortf("FATAL: verdict mismatch")
	}

	// -------------------------------------------------------------------
	// PROCESS PHASE — replay authentication.
	// -------------------------------------------------------------------
	timeout := float64(mustInt(envOr("TOUR_STEP_TIMEOUT", "30")))
	tourRoot := envOr("TOUR_ROOT", filepath.Join(shellOutput(nil, "go", "env", "GOMODCACHE"), "golang.org/x/website@"+tourVersion))
	work, err := os.MkdirTemp("", "tour-evidence-replay")
	if err != nil {
		return abortf("FATAL: %v", err)
	}
	defer os.RemoveAll(work)
	mod := filepath.Join(work, "module")
	if err := evidenceMaterializeModule(mod, tourRoot, items, tcVersion, helperRow); err != nil {
		return abortf("FATAL: cannot materialize replay module: %v", err)
	}
	for _, attempt := range attempts {
		path, mode := asString(attempt["path"]), asString(attempt["mode"])
		item := inventory[path]
		local := filepath.Join(mod, path)
		env := map[string]string{"GOTOOLCHAIN": "local", "BASHY_HINTS": "off"}
		var fresh EvidenceCapture
		if mode == "compiled" {
			compiledDir := filepath.Join(work, "compiled", unsafePathCharRE.ReplaceAllString(path, "_"))
			os.MkdirAll(compiledDir, 0o755)
			pipeline := evidenceCompiledPipeline(bashyPath, goPath, local, mod, compiledDir, timeout, env)
			if pipeline.Stage != asString(attempt["stage"]) {
				return abortf("FATAL: replay pipeline stage mismatch %s %s (recorded %s, replay %s)", path, mode, toS(attempt["stage"]), pipeline.Stage)
			}
			fresh = pipeline.Raw
		} else {
			fresh = evidenceCapture(evidenceCommandFor(mode, item.Applicability, bashyPath, goPath, local), mod, timeout, env)
		}
		freshNormalized := evidenceDerived(fresh.Stdout, fresh.Stderr)
		controlMatch := jsonEqual(fresh.Spawned, attempt["spawned"]) && fresh.State == asString(attempt["state"]) && jsonEqual(fresh.Exit, attempt["exit"])
		if !controlMatch {
			return abortf("FATAL: replay spawn/state/exit not reproduced %s %s (recorded exit=%s state=%s, replay exit=%s state=%s)",
				path, mode, toS(attempt["exit"]), toS(attempt["state"]), toS(fresh.Exit), fresh.State)
		}
		bytesMatch := true
		for _, stream := range []string{"stdout", "stderr"} {
			if !jsonEqual(dig(freshNormalized, stream, "sha256"), dig(attempt, "normalized", stream, "sha256")) {
				bytesMatch = false
			}
		}
		if bytesMatch || mode == "baseline" {
			continue
		}
		return abortf("FATAL: replay raw-output mismatch %s %s — recorded observation was not reproduced by executing its own bound command", path, mode)
	}

	fmt.Printf("Tour evidence VALID %s: 97 programs x 3 modes = 291; root %s; replay-authenticated\n", expectedVerdict, calculatedRoot)
	return 0
}

func expandPath(p string) string {
	abs, err := filepath.Abs(p)
	if err != nil {
		return p
	}
	return abs
}
